internal/component/database_observability/mysql/component.go

```diff
@@ -0,0 +1,260 @@
+	})
+}
+
+var (
+	_ syntax.Defaulter = (*Arguments)(nil)
+	_ syntax.Validator = (*Arguments)(nil)
+)
+
+type Arguments struct {
+	DataSourceName alloytypes.Secret   `alloy:"data_source_name,attr"`
+	ScrapeInterval time.Duration       `alloy:"scrape_interval,attr,optional"`
+	ForwardTo      []loki.LogsReceiver `alloy:"forward_to,attr"`
+}
+
+var DefaultArguments = Arguments{
+	ScrapeInterval: 10 * time.Second,
+}
+
+func (a *Arguments) SetToDefault() {
+	*a = DefaultArguments
+}
+
+func (a *Arguments) Validate() error {
+	_, err := mysql.ParseDSN(string(a.DataSourceName))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type Exports struct {
+	Targets []discovery.Target `alloy:"targets,attr"`
+}
+
+var (
+	_ component.Component    = (*Component)(nil)
+	_ http_service.Component = (*Component)(nil)
+)
+
+type Collector interface {
+	Run(context.Context) error
+	Stop()
+}
+
+type Component struct {
+	opts       component.Options
+	args       Arguments
+	mut        sync.RWMutex
+	receivers  []loki.LogsReceiver
+	handler    loki.LogsReceiver
+	registry   *prometheus.Registry
+	baseTarget discovery.Target
+	collectors []Collector
+}
+
+func New(opts component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:      opts,
+		args:      args,
+		receivers: args.ForwardTo,
+		handler:   loki.NewLogsReceiver(),
+		registry:  prometheus.NewRegistry(),
+	}
+
+	baseTarget, err := c.getBaseTarget()
+	if err != nil {
+		return nil, err
+	}
+	c.baseTarget = baseTarget
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		level.Info(c.opts.Logger).Log("msg", name+" component shutting down, stopping collectors")
+		c.mut.RLock()
+		for _, collector := range c.collectors {
+			collector.Stop()
+		}
+		c.mut.RUnlock()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler.Chan():
+			c.mut.RLock()
+			for _, receiver := range c.receivers {
+				receiver.Chan() <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+func (c *Component) getBaseTarget() (discovery.Target, error) {
+	data, err := c.opts.GetServiceData(http_service.ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get HTTP information: %w", err)
+	}
+	httpData := data.(http_service.Data)
+
+	return discovery.Target{
+		model.AddressLabel:     httpData.MemoryListenAddr,
+		model.SchemeLabel:      "http",
+		model.MetricsPathLabel: path.Join(httpData.HTTPPathForComponent(c.opts.ID), "metrics"),
+		"instance":             c.instanceKey(),
+		"job":                  "integrations/db-o11y",
+	}, nil
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	c.opts.OnStateChange(Exports{
+		Targets: []discovery.Target{c.baseTarget},
+	})
+
+	for _, collector := range c.collectors {
+		collector.Stop()
+	}
+	c.collectors = nil
+
+	newArgs := args.(Arguments)
+	c.args = newArgs
```

Comment:
I think that you can just do c.args := args.(Arguments). newArgs could be useful if you would want the component to keep running with the previous config in case the update goes wrong. In this case you would set the c.args = newArgs at the end of the Update when you are sure that it went through. Since that you are stopping the collectors anyway I don't think that you need this

Reply:
Makes sense! I think it might be a nice followup to keep the component working until the new config is applied, but I'll leave it out for now.